driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -86,12 +92,16 @@
     private final ConcurrentPool<UsageTrackingInternalConnection> pool;
     private final ConnectionPoolSettings settings;
     private final AtomicInteger generation = new AtomicInteger(0);
-    private final AtomicInteger lastPrunedGeneration = new AtomicInteger(0);
     private final ScheduledExecutorService sizeMaintenanceTimer;
     private ExecutorService asyncGetter;
     private final Runnable maintenanceTask;
     private final ConnectionPoolListener connectionPoolListener;
     private final ServerId serverId;
+    private final AtomicInteger numPinnedToCursor = new AtomicInteger(0);
+    private final AtomicInteger numPinnedToTransaction = new AtomicInteger(0);
+
+    private final Map<ObjectId, ServerStats> serverStatsMap = new HashMap<>();
```

Comment:
I propose moving this map together with `ServerStats` to a different class, e.g.
```java
class ServiceStateManager {
    Map<ObjectId, ServerStats> stateByServiceId = ...;

    void addConnection(ObjectId serviceId) { ... }

    void removeConnection(ObjectId serviceId) { ... }

    boolean incrementGeneration(ObjectId serviceId, int expectedGeneration) { ... }

    int getGeneration(ObjectId serviceId) { ... }

    class ServerStats { ... }
}
```
This way we isolate both state and logic related to pool generations specific to services to a separate class.